### ISSUE TITLE
changeData works perfect on html template, can't get it to work calling from controller 

### ISSUE THREAD

**Issue by luisdev94** [l-lin/angular-datatables#952]
- angular version: 1.5.10
- jquery version: 3.1.1
- datatables version: 1.10.13
- angular-datatables version: 0.5.6

I have an angular-datatable with elements. These elements are editable. When an element is edited the table should update accordingly. 

Right now, when the element is edited the table isn't updated automatically, but when I click on a button in the html the table is refreshed. Clearly I'm missing something in the controller to make changeData() work.

html:
```html
<div ng-controller="CompCtrl">
    <div>
        <button class="btn btn-default pull-left" style="background-color: #209e91; color: white;" data-toggle="modal"
         ng-click="open('app/pages/comps/compModal.html', 'md')">
        <span class="glyphicon glyphicon-plus"></span> Add Component</button> &nbsp;
        <button class="btn btn-primary" ng-click="dtInstance.changeData(newPromise)">Refresh</button>
    </div>
    <br><br>
    <div>
        <table datatable="" dt-options="dtOptions" dt-columns="dtColumns" dt-instance="dtInstanceCallback" class="row-border hover"></table>
    </div>
</div>
```
js:
```js
          $scope.newPromise = function () {
            var defer = $q.defer();
            $http.get(url+"component").then(function(result) {
              defer.resolve(result.data.type);
            });
            return defer.promise;
          }

          $scope.dtOptions = DTOptionsBuilder.fromFnPromise($scope.newPromise)
            .withPaginationType('full_numbers');

          $scope.dtColumns = [
            DTColumnBuilder.newColumn('id').withTitle('Id'),
            DTColumnBuilder.newColumn('type').withTitle('Type'),
            DTColumnBuilder.newColumn('name').withTitle('Name'),
            DTColumnBuilder.newColumn('version').withTitle('Version'),
            DTColumnBuilder.newColumn('description').withTitle('Description'),
            DTColumnBuilder.newColumn(null).withTitle('')
              .notSortable()
              .renderWith(function(data, type, full, meta) {
                return '<button class="edit" id="'+data.id+'" onclick="angular.element(this).scope().setId(\''+data.id+'\', \''+data.type+'\', \''+data.name+'\', \''+data.version+'\', \''+data.description+'\')"></button>';
              })
          ];

          $scope.dtInstance = {}

          $scope.dtInstanceCallback = function (instance) {
            $scope.dtInstance = instance;
          }

          $scope.update = function (name, version, description) {
            $scope.demo.name = name;
            $scope.demo.version = version;
            $scope.demo.description = description;
            $http.post(url+"component", $scope.demo)
              .success(function(data) {
                if (data.status==0) {
                  //$scope.dtInstance.changeData($scope.newPromise);
                  if(!angular.equals($scope.dtInstance, {})){
                    $scope.dtInstance.changeData($scope.newPromise);
                    alert("Success");
                  }
                  alert('Gateway updated succesfully no update');
                }
                else {
                  //alert(data.status+" "+$scope.demo.id+" "+$scope.demo.description);
                }
              })
              .error(function(data, status, headers, config){
                alert('Fail');
              });
          }
```
In $scope.update the call to the function:
`$scope.dtInstance.changeData($scope.newPromise)`
isn't called because $scope.dtInstance is {}

I got this from another issues related to this one ([https://github.com/l-lin/angular-datatables/issues/365](url), [https://www.bountysource.com/issues/36170995-can-angular-datatable-promise-be-programmatically-changed](url) ) where the suggested solution was to use dtInstance as a function instead of a variable. I tried that but still can't get it to work.

Hopefully you can spot my mistake.

**Comment by l-lin** [comment]
>  To use this functionality, you must provide a function that returns a promise. Just a promise is not enough.

Try with a function that returns a promise instead:

```js
$scope.dtOptions = DTOptionsBuilder.fromFnPromise(function() {
  return $scope.newPromise;
})
```

**Comment by luisdev94** [comment]
> Try with a function that returns a promise instead: 
> ``` 
> $scope.dtOptions = DTOptionsBuilder.fromFnPromise(function() {
>  return $scope.newPromise;
> }) 

Thanks for the answer.
I tried what you suggested but the table never loads.
I get: "Error: _loadedPromise is undefined" on the browser console.
I think syntax in that part is correct, since I am indeed providing a function that returns a promise although the name of the function could me misleading sorry for that:
```
$scope.newPromise = function () {
            var defer = $q.defer();
            $http.get(url+"component").then(function(result) {
              defer.resolve(result.data.type);
            });
            return defer.promise;
          }
```
I think the problem has to do with dtInstance not being instantiated at the moment the code gets here:
```
if(!angular.equals($scope.dtInstance, {})){
   $scope.dtInstance.changeData($scope.newPromise);
   alert("Success");
}
```
The code inside that 'if' block never runs because dtInstance is equal to {}.
If I remove the 'if' and try to run changeData() I'll get:
"Error: $scope.dtInstance.changeData(..) is not a function"

The exact behavior I'm getting is this:
Initially the table renders perfectly. If I try to add/edit a row on the table, the value is added/edited server-side correctly, but the table doesn't reload automatically. However when I reload through a button in the HTML template the table does reload accordingly. 

I'll keep trying to solve it, please let me know if you think of something else that could be causing this.
Thanks for your help.

**Comment by l-lin** [comment]
Can you edit this [pen](http://codepen.io/l-lin/pen/RKxQZZ?editors=1010) to reproduce your issue?

**Comment by luisdev94** [comment]
Finally found what I was doing wrong. Knew I was missing something silly.
Originally I had this:
```
$scope.dtInstance = {}

$scope.dtInstanceCallback = function (instance) {
  $scope.dtInstance = instance;
}
```
Trying random solutions I commented first line to get this:
```
//$scope.dtInstance = {}

$scope.dtInstanceCallback = function (instance) {
  $scope.dtInstance = instance;
}
```
So I removed the `$scope.dtInstance = {}` and know it works like a charm.
I knew the problem was around here but I didn't know exactly what it was. 
Sorry for taking your time and I hope this can help others.
